src/org/kms/core/MKRequester.java

```diff
@@ -95,13 +99,36 @@ public String getMasterKey() {
 			
 		} catch (ErrorResponse error) {
 			if (error.getErroCode() == 404){
+				String uid = getUserIdFromRequestMK();
 				try {
-					master_key = generateRawMasterKey();
-					String encrypted_master_key = crypto.encryptKey(master_key, crypto.getGlobalKey());
-					if (putMasterKeyToBackend(encrypted_master_key)){
-						decrypted_master_key = master_key;
+					if (!KMSUtils.putWithLock(uid)) {
+						while (KMSUtils.getMap().contains(uid)){
+							Thread.sleep(50);
+						}
+						Response res = conn.downloadObject(DSS_MK_BUCKET, mkObjectName);
+						if (res.getStatusCode() == 200) {
+							master_key = res.getXMLString();
+							if (master_key != null){
+								try {
+									decrypted_master_key = crypto.decryptText(master_key, crypto.getGlobalKey());
+								} catch (Exception e){
+									return null;
+								}
+							}
+						}
+					} else {
+						master_key = generateRawMasterKey();
+						String encrypted_master_key = crypto.encryptKey(master_key, crypto.getGlobalKey());
+						if (putMasterKeyToBackend(encrypted_master_key)){
+							decrypted_master_key = master_key;
+						}
+						KMSUtils.getMap().remove(uid);
```

Comment:
this may lead to a race condition. write access to this map should be serialized